### ISSUE TITLE
support sharded format for downsampling output

### ISSUE THREAD

**Issue by chrisroat** [seung-lab/igneous#99]
I just started thinking about the sharded format, and my initial test with the downsampling showed igneous supported reading the sharded format -- yay!.  The downsampled output is unsharded, which is likely just fine (sharding is most important for mip=0).

Is it technically feasible to automatically determine sharding specs for higher mips and write out sharded format?  For example, one could keep the same #minishard/chunk and #minishard/shard.

**Comment by william-silversmith** [comment]
Yes, this is certainly feasible. I've written the sharded format reader in CloudVolume, but I haven't yet tried to write sharded output for images. Currently, Igneous only supports writing the sharded format for skeletons and graphene meshes. I'll add support for sharded images when possible but my todo list is currently pretty long. I'm currently working on shard support for meshes. PRs are appreciated!

**Comment by chrisroat** [comment]
This issue list is not your todo list, and don't make this a top priority.  It isn't a high priority yet for me, either.  I was just curious if it was possible, or if there was any technical limitation.

**Comment by william-silversmith** [comment]
No worries, it's technically feasible and a good idea. Don't stop posting
issues on my account. 😸

**Comment by william-silversmith** [comment]
Sharded format writing is now integrated into Igneous, but downsampling is still under development.

**Comment by chrisroat** [comment]
I'm looking into this feature, and had a few questions.

The following comment indicates only factor=2x2x1 is supported for sharded downsampling, but I didn't see where that limitation comes from or is enforced.
https://github.com/seung-lab/igneous/blob/02c69224eb39d6590bd6fe967e08288227e92640/igneous/task_creation/image.py#L514

I see that memory constraints requirements require that only a single MIP be calculated at a time (which is enforced) -- is this related?  Is it that only a single 2x2x1 MIP is possible, memory-wise?

My pipeline hands off downsampling tasks to a queue, so I cannot really enforce doing sequential sharded downsamples (unless my worker pool has a single worker).  So for the same memory reasons, I surmise that I can't simply add k downsample tasks, each with factors 2^k, correct?

**Comment by william-silversmith** [comment]
Hi Chris,

Yes, the reason is that the memory blows up quite quickly (can be gigabytes per shard). To keep memory under control, I do a trick where I only download the image regions needed for a thin Z slice at a time, downsample it, and put the resultant downsamples into a shard file. However, if we did even two levels at a time, I would need to keep 5 shards worth in memory. For three levels, 21 shards, and so on. For 2x2x2, the situation worsens. One level might be doable, but two levels would require keeping 9 shards in memory, and three would be 73.

One way to compromise would be to create a sharded mip 0 and do unsharded downsampling above that. That would be almost a 4x reduction in the number of total files which is probably sufficient for many uses.

If you can think of a better approach, let me know!

**Comment by william-silversmith** [comment]
> My pipeline hands off downsampling tasks to a queue, so I cannot really enforce doing sequential sharded downsamples (unless my worker pool has a single worker). So for the same memory reasons, I surmise that I can't simply add k downsample tasks, each with factors 2^k, correct?

Unfortunately, we don't currently have a good way to represent dependencies....

**Comment by chrisroat** [comment]
> One way to compromise would be to create a sharded mip 0 and do unsharded downsampling above that. That would be almost a 4x reduction in the number of total files which is probably sufficient for many uses.

Yep!  That's what I'm currently doing.  It does help with file counts.
